### PR TITLE
Fix for deprecation warnings concerning clipping in Rayleigh part

### PR DESCRIPTION
**Pull request by adybbroe**
Fix for deprecation warnings concerning clipping in Rayleigh part, of angles and redband reflectance inputs

Should fix this warnings which have been there for a while.

Also, the tests made use of the word `blueband`. This is actually a bit confusing, since we recommend using a redband reflection for adjusting the correction in presence of clouds. So, this PR change this to using the `redband_refl` instead.

```
....miniconda3/envs/python38/lib/python3.8/site-packages/numpy/core/fromnumeric.py:58: DeprecationWarning: Passing `np.nan` to mean no clipping in np.clip has always been unreliable, and is now deprecated. In future, this will always return nan, like it already does when min or max are arrays that contain nan. To skip a bound, pass either None or an np.inf of an appropriate sign.
    return bound(*args, **kwds)
```

 - [ ] Closes #xxxx 
 - [x] Tests added 
 - [x] Tests passed: Passes ``pytest pyspectral`` 
 - [x] Passes ``flake8 pyspectral`` 
 - [ ] Fully documented

### CODE REVIEW

File: pyspectral/rayleigh.py [hunk 1]
```diff
@@ -260,10 +260,12 @@ def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=Non
             if redband is not None:
                 redband = from_array(redband, chunks=redband.shape)
 
-        clip_angle = rad2deg(arccos(1. / sunz_sec_coord.max()))
+        clip_angle = nan_to_num(rad2deg(arccos(1. / sunz_sec_coord.max())))
```
Comment (mraspaud) [inline pyspectral/rayleigh.py:263]: you have duplicated lines here, maybe a small function would be nice? Maybe even the docstring of that function could say that the angles are filled with 0s?

File: pyspectral/tests/test_rayleigh.py [hunk 2]
```diff
@@ -188,19 +227,57 @@ def test_get_reflectance(self, get_effective_wvl,
...
+
+        sun_zenith = da.from_array(np.array([67., 32.]), chunks=2)
+        sat_zenith = da.from_array(np.array([45., 18.]), chunks=2)
+        azidiff = da.from_array(np.array([150., 110.]), chunks=2)
+        redband_refl = da.from_array(np.array([108., -0.5]), chunks=2)
+        retv = self.viirs_rayleigh.get_reflectance(sun_zenith, sat_zenith, azidiff, 'M2', redband_refl)
+        self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT4))
+        self.assertIsInstance(retv, da.Array)
+
+        redband_refl = da.from_array(np.array([100., 20.0]), chunks=2)
```
Comment (mraspaud) [inline pyspectral/tests/test_rayleigh.py:276]: Should this be a separate test? It's cleaner when unit tests only test one thing

File: pyspectral/tests/test_rayleigh.py [hunk 3]
```diff
@@ -188,19 +227,57 @@ def test_get_reflectance(self, get_effective_wvl,
...
+        exists.return_code = True
+        get_rsr_version.return_code = RSR_DATA_VERSION
+        get_effective_wvl.return_value = self.cwvl
+
+        sun_zenith = da.from_array(np.array([67., 32.]), chunks=2)
+        sat_zenith = da.from_array(np.array([45., 18.]), chunks=2)
+        azidiff = da.from_array(np.array([150., 110.]), chunks=2)
+        redband_refl = da.from_array(np.array([108., -0.5]), chunks=2)
+        retv = self.viirs_rayleigh.get_reflectance(sun_zenith, sat_zenith, azidiff, 'M2', redband_refl)
+        self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT4))
```
Comment (mraspaud) [inline pyspectral/tests/test_rayleigh.py:273]: I can recommend using `np.testing.assert_allclose`

**Review by mraspaud — COMMENTED** [review]
Thanks for this PR! I have a few inline comments.
Moreover, `nan_to_num` is now used, but I don't see any tests exercising this functionality. Or am I missing something?

File: pyspectral/rayleigh.py (hunk 1 above)
Comment (adybbroe) [reply pyspectral/rayleigh.py:263]: Yes, good point. I have a proposal coming.

**Comment by adybbroe** [comment]
> Thanks for this PR! I have a few inline comments. Moreover, `nan_to_num` is now used, but I don't see any tests exercising this functionality. Or am I missing something?

I thought that was tested but it seems I only tested this functionality for the redband in `test_get_reflectance_dask_redband_outside_clip`... I will have another look at it

File: pyspectral/tests/test_rayleigh.py (hunk 2 above)
Comment (adybbroe) [reply pyspectral/tests/test_rayleigh.py:276]: Hmm, it is part of the same test I would say, namely that if you are outside the range 20-100 then all values give the same result...

File: pyspectral/tests/test_rayleigh.py [hunk 4]
```diff
@@ -163,6 +164,44 @@ def test_rayleigh_init(self, download_luts, exists):
...
+                                                  get_rsr_version, get_reflectance_lut_from_file,
+                                                  download_luts, exists):
+        """Test getting the reflectance correction - using red band reflections outside 0-100."""
+        rayl = TEST_RAYLEIGH_LUT
+        wvl_coord = TEST_RAYLEIGH_WVL_COORD
+        azid_coord = TEST_RAYLEIGH_AZID_COORD
+        sunz_sec_coord = TEST_RAYLEIGH_SUNZ_COORD
+        satz_sec_coord = TEST_RAYLEIGH_SATZ_COORD
+
+        get_reflectance_lut_from_file.return_value = (rayl, wvl_coord, azid_coord,
```
Comment (pnuu) [inline pyspectral/tests/test_rayleigh.py:183]: Use the global constants directly here, no need to lengthen the test by using additional variables.

File: pyspectral/tests/test_rayleigh.py [hunk 5]
```diff
@@ -188,19 +227,57 @@ def test_get_reflectance(self, get_effective_wvl,
...
+    @patch('pyspectral.rayleigh.get_reflectance_lut_from_file')
+    @patch('pyspectral.rsr_reader.RelativeSpectralResponse.'
+           '_get_rsr_data_version')
+    @patch('pyspectral.rayleigh.Rayleigh.get_effective_wavelength')
+    def test_get_reflectance_dask_redband_outside_clip(self, get_effective_wvl,
+                                                       get_rsr_version, get_reflectance_lut_from_file,
+                                                       download_luts, exists):
+        """Test getting the reflectance correction with dask inputs - using red band reflections outside 0-100."""
+        rayl = da.from_array(TEST_RAYLEIGH_LUT, chunks=(10, 10, 10, 10))
+        wvl_coord = TEST_RAYLEIGH_WVL_COORD
```
Comment (pnuu) [inline pyspectral/tests/test_rayleigh.py:254]: Use the contant directly in the call.

File: pyspectral/tests/test_rayleigh.py (hunk 2 above)
Comment (pnuu) [reply pyspectral/tests/test_rayleigh.py:276]: I it isbjust testing values outside thr valid range, then put all those values to the same call. If the point is to check the exact edges, then I'd say it is a separate test. So three test ranges: within valid range, outside valid range, the edge casea.

File: pyspectral/tests/test_rayleigh.py [hunk 6]
```diff
@@ -229,16 +306,16 @@ def test_get_reflectance_dask(self, get_effective_wvl,
         sun_zenith = da.from_array(np.array([67., 32.]), chunks=2)
         sat_zenith = da.from_array(np.array([45., 18.]), chunks=2)
         azidiff = da.from_array(np.array([150., 110.]), chunks=2)
-        blueband = da.from_array(np.array([14., 5.]), chunks=2)
-        retv = self.viirs_rayleigh.get_reflectance(sun_zenith, sat_zenith, azidiff, 'M2', blueband)
+        redband_refl = da.from_array(np.array([14., 5.]), chunks=2)
```
Comment (pnuu) [inline pyspectral/tests/test_rayleigh.py:309]: Any reason why not using `da.array()` directly? Same is used many places, so would simplify the code a lot.

File: pyspectral/tests/test_rayleigh.py (hunk 3 above)
Comment (adybbroe) [reply pyspectral/tests/test_rayleigh.py]: I changed the testing accordingly, thanks!

File: pyspectral/tests/test_rayleigh.py (hunk 4 above)
Comment (adybbroe) [reply pyspectral/tests/test_rayleigh.py]: @pnuu In the mean time I have collected everything in `setUp`, so it should be a lot better now at least.

File: pyspectral/tests/test_rayleigh.py (hunk 5 above)
Comment (adybbroe) [reply pyspectral/tests/test_rayleigh.py]: See comment/answer above.

File: pyspectral/rayleigh.py (hunk 1 above)
Comment (adybbroe) [reply pyspectral/rayleigh.py]: Okay now @mraspaud ?

File: pyspectral/tests/test_rayleigh.py (hunk 6 above)
Comment (adybbroe) [reply pyspectral/tests/test_rayleigh.py]: No idea actually, don't think I did those. But I can change,you are right. Can you actually set the `chunksize` that way? Not that it should matter in this case though...

File: pyspectral/rayleigh.py [hunk 7]
```diff
@@ -236,6 +236,20 @@ def _do_interp(minterp, sunzsec, azidiff, satzsec):
         res = minterp(interp_points2)
         return res.reshape(sunzsec.shape)
 
+    def _clip_angles_inside_coordinate_range(self, zenith_angle, zenith_secant_max):
+        """Clipping solar- or satellite zenith angles to be inside the allowed coordinate range.
+
+        The "allowed" coordinate range is 0 to the angle corresponiding to a
+        maximum secant. The maximum secant is here the maximum available in the
```
Comment (mraspaud) [inline pyspectral/rayleigh.py:243]:
```suggestion
        maximum secant. The maximum secant is where the maximum available in the
```

File: pyspectral/rayleigh.py [hunk 8]
```diff
@@ -260,15 +274,15 @@ def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=Non
...
-        clip_angle = rad2deg(arccos(1. / satz_sec_coord.max()))
-        sat_zenith = clip(sat_zenith, 0, clip_angle)
+
+        sat_zenith = self._clip_angles_inside_coordinate_range(sat_zenith, satz_sec_coord.max())
         satzsec = 1. / cos(deg2rad(sat_zenith))
+
         shape = sun_zenith.shape
 
-        if not(wvl_coord.min() < wvl < wvl_coord.max()):
+        if not wvl_coord.min() < wvl < wvl_coord.max():
```
Comment (mraspaud) [inline pyspectral/rayleigh.py:285]: Should this be done first in the function, so that the lut doesn't need to be read at all if the wavelength is outside the visible range?

Comment (adybbroe) [reply pyspectral/rayleigh.py:285]: Well, that would be nice, but things are quite entangled at the moment with the Dask / non-Dask code. Before returning the zero-array a chunk size is being calculated from the sun-zenith and redband arrays, so they are needed here atm! :-(

There is a need for refactoring indeed. But rather do that in another PR

**Review by mraspaud — APPROVED** [review]
LGTM
